# Introduction to Pandas

## Pandas provides Python data frames

* Popular and established
* Inspired by R dataframes
* Built on `numpy` for fast computation

In [1]:
import pandas as pd

## Our first dataframe

In [2]:
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "Love_of_R": [2, 5, 11],
                   "years_at_wsu": [4, 17, 5]})
df.head()

Names  Python_mastery  Love_of_R  years_at_wsu
0  Iverson            10.0          2             4
1   Malone             5.0          5            17
2   Bergen             1.0         11             5

## Reading from a csv

* Most data sets will be read in from a csv or JSON data file
* `Pandas` provides `read_csv` and `read_json`

### Open a local file w/ relative path

In [3]:
# Won't work in colab
artists = pd.read_csv('./data/Artists.csv')
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

### Open a web address

In [4]:
url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.csv"
artists =  pd.read_csv(url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992       NaN          NaN  
1       1936        0       NaN          NaN  
2       1941        0       NaN          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0       NaN          NaN

# JSON data file

* Another (more modern) storage
* Here the data is stored in row `dict`

```{json}
[
{
  "ConstituentID": 1,
  "DisplayName": "Robert Arneson",
  "ArtistBio": "American, 1930–1992",
  "Nationality": "American",
  "Gender": "Male",
  "BeginDate": 1930,
  "EndDate": 1992,
  "Wiki QID": null,
  "ULAN": null
},
{
  "ConstituentID": 2,
  "DisplayName": "Doroteo Arnaiz",
  "ArtistBio": "Spanish, born 1936",
  "Nationality": "Spanish",
  "Gender": "Male",
  "BeginDate": 1936,
  "EndDate": 0,
  "Wiki QID": null,
  "ULAN": null
},
...
```

## `pandas` can read `json` data

In [5]:
# Won't work in colab
artists =  pd.read_json('./data/Artists.json')
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

In [6]:
json_url = "https://github.com/MuseumofModernArt/collection/raw/master/Artists.json"
artists =  pd.read_json(json_url)
artists.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID         ULAN  
0       1930     1992      None          NaN  
1       1936        0      None          NaN  
2       1941        0      None          NaN  
3       1946        0  Q1063584  500027998.0  
4       1941        0      None          NaN

## <font color="red"> Exercise 2.1.2 </font>
    
Use tab-completion and `help` to discover and explore two more methods of reading a file into a `Pandas` dataframe.


In [7]:
help(pd.read_clipboard)

Help on function read_clipboard in module pandas.io.clipboards:

read_clipboard(sep: 'str' = '\\s+', **kwargs)
    Read text from clipboard and pass to read_csv.
    
    Parameters
    ----------
    sep : str, default '\s+'
        A string or regex delimiter. The default of '\s+' denotes
        one or more whitespace characters.
    
    **kwargs
        See read_csv for the full argument list.
    
    Returns
    -------
    DataFrame
        A parsed DataFrame object.



In [8]:
help(pd.read_feather)

Help on function read_feather in module pandas.io.feather_format:

read_feather(path: 'FilePath | ReadBuffer[bytes]', columns: 'Sequence[Hashable] | None' = None, use_threads: 'bool' = True, storage_options: 'StorageOptions' = None)
    Load a feather-format object from the file path.
    
    Parameters
    ----------
    path : str, path object, or file-like object
        String, path object (implementing ``os.PathLike[str]``), or file-like
        object implementing a binary ``read()`` function. The string could be a URL.
        Valid URL schemes include http, ftp, s3, and file. For file URLs, a host is
        expected. A local file could be: ``file://localhost/path/to/table.feather``.
    columns : sequence, default None
        If not provided, all columns are read.
    use_threads : bool, default True
        Whether to parallelize reading using multiple threads.
    storage_options : dict, optional
        Extra options that make sense for a particular storage connection, e.

> Discuss what you found here: Pandas can read just about any file format (and if there was a format it couldn't read, there are almost certainly other packages that can.

## <font color="red"> Exercise 2.1.2 </font>
    
Read in the `Artwork.csv` from [https://github.com/MuseumofModernArt/collection](https://github.com/MuseumofModernArt/collection) and display the head of the resulting dataframe.


In [9]:
pd.set_option("display.max_columns", None) # show all the columns when we look at the data

In [10]:
# Note: going to the page on GH will have blob instead of raw in the URL due to file size. Using that results in a ParseError
artwork = pd.read_csv("https://github.com/MuseumofModernArt/collection/raw/master/Artworks.csv") # works fine
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper   
1  (Male)  1987                  Paint and colored pencil on print   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...   
3  (Male)  1980  Photographic reproduction with colored synthet...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...   

                           Dimensions  \
0  19 1/8 x 66 1/2" (48.6 x 168.9 cm)   
1       16 x 11 3/4" (40.6 x 29.8 cm)   
2   13 1/2 x 12 1/2" (34.3 x 31.8 cm)   
3           20 x 20" (50.8 x 50.8 cm)   
4    15 1/8 x 7 1/2" (38.4 x 19.1 cm)   

                                          CreditLine AccessionNumber  \
0  Fractional and promised gift of Jo Carole and ...        885.1996   
1  Gift of the architect in honor of Lily Auchinc...          1.1995   
2             Gift of Jo Carole and Ronald S. Lauder          1.1997   
3  Purchase and partial gift of the architect in ...          2.1995   
4             Gift of Jo Carole and Ronald S. Lauder          2.1997   

  Classification             Department DateAcquired Cataloged  ObjectID  \
0   Architecture  Architecture & Design   1996-04-09         Y         2   
1   Architecture  Architecture & Design   1995-01-17         Y         3   
2   Architecture  Architecture & Design   1997-01-15         Y         4   
3   Architecture  Architecture & Design   1995-01-17         Y         5   
4   Architecture  Architecture & Design   1997-01-15         Y         6   

                                      URL  \
0  http://www.moma.org/collection/works/2   
1  http://www.moma.org/collection/works/3   
2  http://www.moma.org/collection/works/4   
3  http://www.moma.org/collection/works/5   
4  http://www.moma.org/collection/works/6   

                                        ThumbnailURL  Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                 NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                 NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                 NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                 NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                 NaN   

   Depth (cm)  Diameter (cm)  Height (cm)  Length (cm)  Weight (kg)  \
0         NaN            NaN      48.6000          NaN          NaN   
1         NaN            NaN      40.6401          NaN          NaN   
2         NaN            NaN      34.3000          NaN          NaN   
3         NaN            NaN      50.8000          NaN          NaN   
4         NaN            NaN      38.4000          NaN          NaN   

   Width (cm)  Seat Height (cm)  Duration (sec.)  
0    168.9000               NaN              NaN  
1     29.8451               NaN              NaN  
2     31.8000               NaN              NaN  
3     50.8000               NaN              NaN  
4 

## So what is a `DateFrame`

* Like R, Pandas focuses on columns
* Think `dict` of `(str, Series)` pairs 
* A series is a typed list-like structure

In [11]:
# This is how I imagine a dataframe
df = pd.DataFrame({"Names": ["Iverson", "Malone", "Bergen"],
                   "Python_mastery": [10, 5, 1.0],
                   "years_at_wsu": [4.5, 17.5, 5.5]})

In [12]:
type(df)

pandas.core.frame.DataFrame

## Columns are `Series` and hold one type of data

In [13]:
type(artists.BeginDate), type(artists.DisplayName)

(pandas.core.series.Series, pandas.core.series.Series)

In [14]:
artists.BeginDate.dtype, artists.DisplayName.dtype

(dtype('int64'), dtype('O'))

## Two ways to access a column

* **Method 1:** like a dictionary
    * `df["column_name"]`
* **Method 2:** like an object attribute
    * `df.column_name`
    * Only for proper names!

In [15]:
artists.BeginDate.head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

In [16]:
artists['BeginDate'].head(2)

0    1930
1    1936
Name: BeginDate, dtype: int64

## More on data types

* See all data types with `df.dtypes`
* You can set the `dtypes` when you read a dataframe
* Read more about types: [Pandas docs](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dtypes.html)

In [17]:
artists.dtypes

ConstituentID      int64
DisplayName       object
ArtistBio         object
Nationality       object
Gender            object
BeginDate          int64
EndDate            int64
Wiki QID          object
ULAN             float64
dtype: object

## Setting `dtypes` with `read_csv`

We can pass a `dict` of types to `dtype` keyword

In [18]:
import numpy as np
artist_types = {'ConstituentID': np.int64,
                'DisplayName': str,
                'ArtistBio': str,
                'Nationality': str,
                'Gender':str,
                'BeginDate': np.int64,
                'EndDate': np.int64,
                'Wiki QID': str,
                'ULAN':pd.Int64Dtype()} # If you get an error ==> update pandas (see below)
artists2 = pd.read_csv('./data/Artists.csv', dtype = artist_types)
artists2.head()

ConstituentID      DisplayName            ArtistBio Nationality Gender  \
0              1   Robert Arneson  American, 1930–1992    American   Male   
1              2   Doroteo Arnaiz   Spanish, born 1936     Spanish   Male   
2              3      Bill Arnold  American, born 1941    American   Male   
3              4  Charles Arnoldi  American, born 1946    American   Male   
4              5      Per Arnoldi    Danish, born 1941      Danish   Male   

   BeginDate  EndDate  Wiki QID       ULAN  
0       1930     1992       NaN       <NA>  
1       1936        0       NaN       <NA>  
2       1941        0       NaN       <NA>  
3       1946        0  Q1063584  500027998  
4       1941        0       NaN       <NA>

## What's up with `ULAN` ?

* Currently, `numpy` $\rightarrow$ no missing `int`s
* Pandas correct this with the `pd.Int64Dtype()` type
    * Only available in `pandas >= 0.24.0`

In [19]:
pd.__version__

'1.4.3'

## An `Int` by any other name ...

* `np.int64` $\rightarrow$ no missing values
* `pd.Int64Dtype()` $\rightarrow$ allows `NaN`

In [20]:
artists2.dtypes

ConstituentID     int64
DisplayName      object
ArtistBio        object
Nationality      object
Gender           object
BeginDate         int64
EndDate           int64
Wiki QID         object
ULAN              Int64
dtype: object

## Preview of coming attractions

* Now we can switch `BeginDate` and `EndDate` from `0` to `np.NaN`
* We will do this in the next section

# Getting to know your data

## Basic inspection tools

* `df.head()`        first five rows
* `df.tail()`        last five rows
* `df.sample(5)`     random sample of rows
* `df.shape`         number of rows/columns in a tuple
* `df.describe()`    calculates measures of central tendency
* `df.info()`

## <font color="red"> Exercise 1: Inspect the artwork from MoMA </font>

#### Read the csv and inspect the `head`

In [21]:
artwork.head()

Title  \
0  Ferdinandsbrücke Project, Vienna, Austria (Ele...   
1  City of Music, National Superior Conservatory ...   
2  Villa near Vienna Project, Outside Vienna, Aus...   
3  The Manhattan Transcripts Project, New York, N...   
4  Villa, project, outside Vienna, Austria, Exter...   

                     Artist ConstituentID  \
0               Otto Wagner          6210   
1  Christian de Portzamparc          7470   
2                Emil Hoppe          7605   
3           Bernard Tschumi          7056   
4                Emil Hoppe          7605   

                                   ArtistBio Nationality BeginDate EndDate  \
0                      (Austrian, 1841–1918)  (Austrian)    (1841)  (1918)   
1                        (French, born 1944)    (French)    (1944)     (0)   
2                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   
3  (French and Swiss, born Switzerland 1944)          ()    (1944)     (0)   
4                      (Austrian, 1876–1957)  (Austrian)    (1876)  (1957)   

   Gender  Date                                             Medium  \
0  (Male)  1896      Ink and cut-and-pasted painted pages on paper   
1  (Male)  1987                  Paint and colored pencil on print   
2  (Male)  1903  Graphite, pen, color pencil, ink, and gouache ...   
3  (Male)  1980  Photographic reproduction with colored synthet...   
4  (Male)  1903  Graphite, color pencil, ink, and gouache on tr...   

                           Dimensions  \
0  19 1/8 x 66 1/2" (48.6 x 168.9 cm)   
1       16 x 11 3/4" (40.6 x 29.8 cm)   
2   13 1/2 x 12 1/2" (34.3 x 31.8 cm)   
3           20 x 20" (50.8 x 50.8 cm)   
4    15 1/8 x 7 1/2" (38.4 x 19.1 cm)   

                                          CreditLine AccessionNumber  \
0  Fractional and promised gift of Jo Carole and ...        885.1996   
1  Gift of the architect in honor of Lily Auchinc...          1.1995   
2             Gift of Jo Carole and Ronald S. Lauder          1.1997   
3  Purchase and partial gift of the architect in ...          2.1995   
4             Gift of Jo Carole and Ronald S. Lauder          2.1997   

  Classification             Department DateAcquired Cataloged  ObjectID  \
0   Architecture  Architecture & Design   1996-04-09         Y         2   
1   Architecture  Architecture & Design   1995-01-17         Y         3   
2   Architecture  Architecture & Design   1997-01-15         Y         4   
3   Architecture  Architecture & Design   1995-01-17         Y         5   
4   Architecture  Architecture & Design   1997-01-15         Y         6   

                                      URL  \
0  http://www.moma.org/collection/works/2   
1  http://www.moma.org/collection/works/3   
2  http://www.moma.org/collection/works/4   
3  http://www.moma.org/collection/works/5   
4  http://www.moma.org/collection/works/6   

                                        ThumbnailURL  Circumference (cm)  \
0  http://www.moma.org/media/W1siZiIsIjU5NDA1Il0s...                 NaN   
1  http://www.moma.org/media/W1siZiIsIjk3Il0sWyJw...                 NaN   
2  http://www.moma.org/media/W1siZiIsIjk4Il0sWyJw...                 NaN   
3  http://www.moma.org/media/W1siZiIsIjEyNCJdLFsi...                 NaN   
4  http://www.moma.org/media/W1siZiIsIjEyNiJdLFsi...                 NaN   

   Depth (cm)  Diameter (cm)  Height (cm)  Length (cm)  Weight (kg)  \
0         NaN            NaN      48.6000          NaN          NaN   
1         NaN            NaN      40.6401          NaN          NaN   
2         NaN            NaN      34.3000          NaN          NaN   
3         NaN            NaN      50.8000          NaN          NaN   
4         NaN            NaN      38.4000          NaN          NaN   

   Width (cm)  Seat Height (cm)  Duration (sec.)  
0    168.9000               NaN              NaN  
1     29.8451               NaN              NaN  
2     31.8000               NaN              NaN  
3     50.8000               NaN              NaN  
4 

**Task:** Write a few sentences describing an problems

* Some items, including but not limited to Nationality, BeginDate, and Gender, are wrapped in parentheses.  
* There are missing items in multiple fields which are handled in different ways. Missing nationality = (), missing end date = (0), missing metrics = NaN.  
* The ArtistBio field seems pretty redundant, given we also have Nationality and BeginDate and EndDate fields, although it does seem to sometimes have data that those do not, so we should probably update those fields using ArtistBio before removing it.  
* BeginDate and EndDate could use better column names - I assumed those related to the piece of art, but they are actually the birth and death years of the artist.

#### Inspect the column names with the `columns` attribute

In [22]:
artwork.columns

Index(['Title', 'Artist', 'ConstituentID', 'ArtistBio', 'Nationality',
       'BeginDate', 'EndDate', 'Gender', 'Date', 'Medium', 'Dimensions',
       'CreditLine', 'AccessionNumber', 'Classification', 'Department',
       'DateAcquired', 'Cataloged', 'ObjectID', 'URL', 'ThumbnailURL',
       'Circumference (cm)', 'Depth (cm)', 'Diameter (cm)', 'Height (cm)',
       'Length (cm)', 'Weight (kg)', 'Width (cm)', 'Seat Height (cm)',
       'Duration (sec.)'],
      dtype='object')

**Question:** See any problems?

* Dimensions is probably not a very useful field (especially given the other Length/Width/Height fields) and should be removed.
* Date format could potentially pose challenges if it is not well-structured.

#### Inspect the tail

In [23]:
artwork.tail()

Title           Artist ConstituentID  \
139932  48 Jugendstil postcards  Various Artists          6104   
139933  48 Jugendstil postcards  Various Artists          6104   
139934  48 Jugendstil postcards  Various Artists          6104   
139935  48 Jugendstil postcards  Various Artists          6104   
139936                 Untitled              NaN           NaN   

                    ArtistBio            Nationality BeginDate EndDate Gender  \
139932  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139933  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139934  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139935  (Nationality unknown)  (Nationality unknown)       (0)     (0)     ()   
139936                    NaN                    NaN       NaN     NaN    NaN   

           Date                           Medium  \
139932  c. 1898                              NaN   
139933  c. 1898                              NaN   
139934  c. 1898                              NaN   
139935  c. 1898                              NaN   
139936     1993  Chromogenic print, printed 2012   

                                    Dimensions  \
139932              3 1/2 x 5 1/2" (9 x 14 cm)   
139933              3 1/2 x 5 1/2" (9 x 14 cm)   
139934              3 1/2 x 5 1/2" (9 x 14 cm)   
139935              3 1/2 x 5 1/2" (9 x 14 cm)   
139936  8 11/16 × 12 13/16" (22 × 32.5 cm)\r\n   

                                               CreditLine AccessionNumber  \
139932                          Gift of Leonard A. Lauder     252.1987.45   
139933                          Gift of Leonard A. Lauder     252.1987.46   
139934                          Gift of Leonard A. Lauder     252.1987.47   
139935                          Gift of Leonard A. Lauder     252.1987.48   
139936  The Family of Man Fund, Carol and David Appel ...     CH2022.1890   

       Classification             Department DateAcquired Cataloged  ObjectID  \
139932         Design  Architecture & Design   1987-05-27         N    436514   
139933         Design  Architecture & Design   1987-05-27         N    436515   
139934         Design  Architecture & Design   1987-05-27         N    436516   
139935         Design  Architecture & Design   1987-05-27         N    436517   
139936     Photograph            Photography          NaN         N    436925   

        URL ThumbnailURL  Circumference (cm)  Depth (cm)  Diameter (cm)  \
139932  NaN          NaN                 NaN         NaN            NaN   
139933  NaN          NaN                 NaN         NaN            NaN   
139934  NaN          NaN                 NaN         NaN            NaN   
139935  NaN          NaN                 NaN         NaN            NaN   
139936  NaN          NaN                 NaN         NaN            NaN   

        Height (cm)  Length (cm)  Weight (kg)  Width (cm)  Seat Height (cm)  \
139932          9.0          NaN          NaN        14.0               NaN   
139933          9.0          NaN          NaN        14.0               NaN   
139934          9.0          NaN          NaN        14.0               NaN   
139935          9.0          NaN          NaN        14.0               NaN   
139936          NaN          NaN          NaN         NaN               NaN   

        Duration (sec.)  
139932              NaN  
139933              NaN  
139934              NaN  
139935              NaN  
139936              NaN

#### Check out the `shape`

In [24]:
artwork.shape

(139937, 29)

In [25]:
print(len(artwork.columns))
print(len(artwork))

29
139937


**Question:** What do these number mean?

We have 139,937 rows/observations/cases (no wonder it took a minute to load) of 29 variables.

#### Use `describe` to compute statistics

In [26]:
artwork.describe()

ObjectID  Circumference (cm)    Depth (cm)  Diameter (cm)  \
count  139937.000000           10.000000  15166.000000    1464.000000   
mean   102094.204656           44.868020     15.227029      23.226111   
std     90877.170308           28.631604     52.885909      44.902258   
min         2.000000            9.900000      0.000000       0.635000   
25%     37161.000000           23.500000      0.000000       7.778800   
50%     74919.000000           36.000000      0.000000      13.652500   
75%    144644.000000           71.125000      7.800000      25.000000   
max    436925.000000           83.800000   1808.483617     914.400000   

         Height (cm)  Length (cm)    Weight (kg)     Width (cm)  \
count  122889.000000   742.000000     287.000000  121968.000000   
mean       37.420501    89.687579    1299.039319      37.935016   
std        49.478379   329.428165   12079.491475      67.180817   
min         0.000000     0.000000       0.090000       0.000000   
25%        17.800000    17.100000       5.896800      17.500000   
50%        27.800000    26.700000      20.865500      25.400000   
75%        43.815088    79.700000      80.967100      44.132588   
max      9140.000000  8321.056600  185067.585957    9144.000000   

       Seat Height (cm)  Duration (sec.)  
count               0.0     1.949000e+03  
mean                NaN     6.081282e+03  
std                 NaN     1.434418e+05  
min                 NaN     0.000000e+00  
25%                 NaN     1.200000e+02  
50%                 NaN     4.200000e+02  
75%                 NaN     1.500000e+03  
max                 NaN     6.283065e+06

#### Use `info` to look at types and totals

In [27]:
artwork.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 139937 entries, 0 to 139936
Data columns (total 29 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Title               139898 non-null  object 
 1   Artist              138711 non-null  object 
 2   ConstituentID       138711 non-null  object 
 3   ArtistBio           134091 non-null  object 
 4   Nationality         138711 non-null  object 
 5   BeginDate           138711 non-null  object 
 6   EndDate             138711 non-null  object 
 7   Gender              138711 non-null  object 
 8   Date                137812 non-null  object 
 9   Medium              130213 non-null  object 
 10  Dimensions          130606 non-null  object 
 11  CreditLine          138084 non-null  object 
 12  AccessionNumber     139937 non-null  object 
 13  Classification      139936 non-null  object 
 14  Department          139937 non-null  object 
 15  DateAcquired        133251 non-nul

**Question:** What did you learn from the last two cells?

* Seat Height is a completely useless field as all the values are null.
* Most of the fields are missing some values.
* The type inferrer could use some help as we could probably do better than object for most of the fields.